spec/requests/portfolios_spec.rb

```diff
@@ -440,50 +440,84 @@
-        expect(json['name']).to eq(tag_params[:name])
-        expect(json['namespace']).to eq(tag_params[:namespace]) if tag_params.key?(:namespace)
-        expect(json['value']).to eq(tag_params[:value]) if tag_params.key?(:value)
+        post "#{api}/portfolios/#{portfolio.id}/tag", :headers => default_headers, :params => tag_params
+        expect(json['name']).to eq(params[:name])
+        expect(json['namespace']).to eq(params[:namespace]) if params.key?(:namespace)
+        expect(json['value']).to eq(params[:value]) if params.key?(:value)
         expect(response).to have_http_status(200)
       end
     end
 
     context "GET /portfolios/{id}/tags" do
+      let(:params) { {:name => tag_name, :namespace => tag_ns} }
+
       before do
-        portfolio.tag_add("test_tag")
+        post "#{api}/portfolios/#{portfolio.id}/tag", :headers => default_headers, :params => tag_params
       end
 
       it "returns the tags for the portfolio" do
         get "#{api}/portfolios/#{portfolio.id}/tags", :headers => default_headers
 
         expect(json["meta"]["count"]).to eq 1
-        expect(json["data"].first["name"]).to eq "test_tag"
+        expect(json["data"].first["name"]).to eq tag_name
+      end
+
+      it "allows filtering on the response" do
+        get "#{api}/portfolios/#{portfolio.id}/tags?filter[namespace][eq]=#{tag_ns}&filter[name][eq]=#{tag_name}", :headers => default_headers
+
+        expect(json["meta"]["count"]).to eq 1
+        expect(json["data"].first["name"]).to eq tag_name
       end
     end
 
     context "POST /portfolios/{id}/tag" do
       context 'no namespace and value' do
-        let(:tag_params) { { :name => tag_name } }
+        let(:params) { {:name => tag_name} }
         it_behaves_like "#tag_add_test"
       end
 
       context 'no value' do
-        let(:tag_params) { { :name => tag_name, :namespace => tag_ns } }
+        let(:params) { {:name => tag_name, :namespace => tag_ns} }
         it_behaves_like "#tag_add_test"
       end
 
       context 'all in' do
-        let(:tag_params) { { :name => tag_name, :namespace => tag_ns, :value => tag_value } }
+        let(:params) { {:name => tag_name, :namespace => tag_ns, :value => tag_value} }
         it_behaves_like "#tag_add_test"
       end
+
+      context 'double add tags' do
+        let(:params) { {:name => tag_name} }
+
+        before do
+          post "#{api}/portfolios/#{portfolio.id}/tag", :headers => default_headers, :params => tag_params
+          post "#{api}/portfolios/#{portfolio.id}/tag", :headers => default_headers, :params => tag_params
+        end
+
+        it "returns not modified" do
+          expect(response).to have_http_status(304)
+        end
+      end
+    end
+
+    context "POST /portfolios/{id}/untag" do
+      let(:name) { 'Gnocchi' }
+      let(:params) do
+        {:tag => Tag.new(:name => name).to_tag_string}
+      end
+
+      it "deletes the tag from the portfolio item" do
```

Comment:
Wondering if this should be renamed to  it 'untags' ..

Reply:
Changed to `removes the tag from the portfolio item`